README.md

```diff
@@ -2,9 +2,118 @@
+
+A solution is to encrypt the configuration data, set it in the ENV encrypted, and then allow the application to decrypt when it boots. This library aims to make this pattern easier to adopt.
+
+First off, the app configuration needs to be encrypted. This should ideally be done by an automated tool. For example:
+
+```ruby
+app_config = {
+  "DB_URL" => "postgres://user:password123@thedb.hostname.com:1234/db_name",
+  "CACHE_URL" => "redis://user:password456@another.url.com:5678",
+  "PAYMENTS_GATEWAY_API_TOKEN" => "SECRET_TOKEN_ZOMG",
+}
+
+private_key = File.read("path/to/rsa_key.pem")
+
+File.write("encrypted_env.txt",  CryptoEnvVar.encrypt(app_config, private_key))
+```
+
+Done that, the encrypted configuration and the public key need to be passed to the starting application. By default, `CryptoEnvVar` will try to read them from two ENV variables:
+
+```
+export CRYPTO_ENV='the base64 encrypted configuration generated above'
+export CRYPTO_ENV_DECRYPT_KEY='the public key'
+
+ruby my_app.rb
+```
+
+If you choose the default names, then all you need to do is call `CryptoEnvVar.bootstrap!` early in the application initialization process (for example in `config.ru` for a Rack app), and it will load and decrypt the encrypted env, then copy its data into the `ENV` of the current process.
+
+```ruby
+ENV["DB_URL"]
+# => nil
+
+CryptoEnvVar.bootstrap!
+
+ENV["DB_URL"]
+# => "postgres://user:password123@thedb.hostname.com:1234/db_name"
+```
+
+
+The sources of both the encrypted configuration and the public key can be configured with the `:read_from` and `:decrypt_with` options, respectively. Valid values are strings or callable objects (you can use a proc or lambda or you can implement your own loaders). For example:
+
+```ruby
+CryptoEnvVar.bootstrap!(
+  read_from: -> { File.read(ENV["secrets_file_path"]).chomp },
+  decrypt_with: SecurePublicKeyFetcher.new
+)
+
+class SecurePublicKeyFetcher
+  def call
+    # ...
+  end
+end
+```
+
+
+By default `CryptoEnvVar.bootstrap!` will copy all the decrypted configuration variables into the ENV, overriding any preset value. If you want to disable this, for example because you want any explicitly set ENV variable to have the precence, you can do so with this option:
+
+```ruby
+CryptoEnvVar.bootstrap!(override_env: false)
+```
+
+
+## A note on the RSA keypairs
+
+Only keypairs without passphrase are supported at this stage.
+
+You can create an RSA keypair in a shell with:
+
+```
+openssl genrsa -out private_key.pem 2048
+openssl rsa -pubout -in private_key.pem -out public_key.pem
+```
+
+Please note that the public key is **NOT** the same thing as the SSH public key file [normally generated with `ssh-keygen`](https://help.github.com/articles/generating-a-new-ssh-key-and-adding-it-to-the-ssh-agent/). You can still use a private RSA key generated with `ssh-keygen`, but then you have to extract the public key with the `openssl` command shown above.
```

Comment:
There's also [this helper gem](https://github.com/bensie/sshkey) I'm using in `hopper`.

Reply:
Yes, I saw that, but I was trying to keep the dependencies to a minimum (reason why initially I went with custom string-chaining, rather than `msgpack`).

It can also be done with the `openssl` package from the stdlib:

```ruby
require "openssl"

key = OpenSSL::PKey::RSA.new(source_key_string)

private_key = key.to_s
public_key = key.public_key.to_s

private_key == source_key_string # true
```

I should probably add it to the readme 🤔 .

Reply:
Done.